# Анализ базы данных сервиса для чтения книг по подписке

## Цели исследования

Проанализировать базу данных крупного сервиса для чтения книг по подписке для формирования ценностного предложения для нового продукта.

База данных содержит информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

## Исследование таблиц 

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
# Список таблиц БД, информацию о которых нужно вывести
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']

# Функция для получения первых 5 строк из каждой таблицы 
# Функция возвращает словарь, где ключами являются имена таблиц, а значениями — соответствующие DataFrame
def get_sql_data(tables: list, engine) -> dict:
    '''Получаем первые 5 строк из каждой таблицы в списке'''
    results = {}
    with engine.connect() as con:
        for table in tables:
            query = f'''SELECT * 
                        FROM {table};
                     '''
            df = pd.read_sql(sql=text(query), con=con)
            results[table] = df
    return results

# Получаем данные из всех таблиц
table_data = get_sql_data(tables, engine)

# Выводим данные для каждой таблицы
for table_name, data in table_data.items(): # возвращает все элементы словаря как последовательность кортежей (ключ, значение)
    display (f'{table_name}', data)
    data.info()
    print(' ')
    print(f'В таблице {table_name} {data.shape[0]} строк, пропусков нет')
    print(' ')

'books'

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
 
В таблице books 1000 строк, пропусков нет
 


'authors'

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
 
В таблице authors 636 строк, пропусков нет
 


'publishers'

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
 
В таблице publishers 340 строк, пропусков нет
 


'ratings'

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
 
В таблице ratings 6456 строк, пропусков нет
 


'reviews'

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
 
В таблице reviews 2793 строк, пропусков нет
 


**Вывод:**  
База данных состоит из пяти таблиц:   

- `books` - cодержит данные о книгах, 1000 ненулевых строк, типы данных соответствуют сохраненным в них значениям во всех столбцах, кроме столбца `publication_date` - дата публикации (тип object вместо datatime)  
  
- `authors` - cодержит данные об авторах, 636 ненулевых строк, типы данных соответствуют сохраненным в них значениям.  
  
- `publishers` - cодержит данные об издательствах, 340 ненулевых строк, типы данных соответствуют сохраненным в них значениям.

- `ratings` - cодержит данные о пользовательских оценках книг, 6456 ненулевых строк, типы данных соответствуют сохраненным в них значениям.  
  
- `reviews` - cодержит данные о пользовательских обзорах, 2793 ненулевых строки, типы данных соответствуют сохраненным в них значениям. 

## Анализ базы данных 

### Количество книг, вышедших после 1 января 2000 года

*Посчитайте, сколько книг вышло после 1 января 2000 года.*

In [5]:
query = '''SELECT COUNT(book_id) AS books_cnt
           FROM books 
           WHERE publication_date >= '2000-01-01';
        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

books_cnt
0        821

После 1 января 2000 года вышла **821 книга**.

### Количество обзоров и средняя оценка каждой книги

*Для каждой книги посчитайте количество обзоров и среднюю оценку.*

In [6]:
query = '''WITH
           b AS (SELECT b.book_id,
                        a.author,
                        b.title
                 FROM books AS b LEFT JOIN authors AS a ON b.author_id = a.author_id),
           rat AS (SELECT book_id,
                          ROUND(AVG(rating), 2) AS rating_avg 
                   FROM ratings 
                   GROUP BY book_id),
           rev AS (SELECT book_id,
                          COUNT(review_id) AS review_cnt
                   FROM reviews 
                   GROUP BY book_id )           
           SELECT b.book_id,
                  b.author,
                  b.title,
                  rat.rating_avg,
                  rev.review_cnt 
           FROM b LEFT JOIN rat ON b.book_id = rat.book_id
                  LEFT JOIN rev ON b.book_id = rev.book_id
           ORDER BY rat.rating_avg DESC;
        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                             author  \
0        518                                    James Patterson   
1        732                                     Steven Johnson   
2        347                                      Tamora Pierce   
3        610                                      James Clavell   
4        330                                     Nigella Lawson   
..       ...                                                ...   
995      915                                 Thomas L. Friedman   
996      371  William S. Burroughs/Oliver  Harris/Allen Gins...   
997      316                                    Joseph J. Ellis   
998      202                                  Christina Schwarz   
999      303                                       Jodi Picoult   

                                                 title  rating_avg  review_cnt  
0                 Pop Goes the Weasel (Alex Cross  #5)        5.00         2.0  
1    The Ghost Map: The Story of London's Most Terr...        5.00         2.0  
2    In the Hand of the Goddess (Song of the Liones...        5.00         2.0  
3                             Tai-Pan (Asian Saga  #2)        5.00         2.0  
4    How to Be a Domestic Goddess: Baking and the A...        5.00         1.0  
..                                                 ...         ...         ...  
995  The World Is Flat: A Brief History of the Twen...        2.25         3.0  
996                                              Junky        2.00         2.0  
997                  His Excellency: George Washington        2.00         2.0  
998                                      Drowning Ruth        2.00         3.0  
999                               Harvesting the Heart        1.50         2.0  

[1000 rows x 5 columns]

В сводной таблице представлена информация о каждой книге: автор, название, средняя оценка и количество обзоров.   
Данные отсортированы по убыванию средней оценки книг.  

###  Издательство, которое выпустило наибольшее число книг толще 50 страниц

*Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.*

In [7]:
query = '''SELECT p.publisher,
                  b.book_cnt
           FROM (SELECT publisher_id,
                        COUNT(book_id) AS book_cnt
                 FROM books 
                 WHERE num_pages > 50
                 GROUP BY publisher_id) AS b
           JOIN publishers AS p ON b.publisher_id = p.publisher_id
           ORDER BY b.book_cnt DESC
           LIMIT 1;
        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  book_cnt
0  Penguin Books        42

Издательство **Penguin Books** выпустило 42 книги - максимальное число книг толще 50 страниц.

### Автор с самой высокой средней оценкой книг

*Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.*

In [9]:
query = '''WITH
           b AS (SELECT b.book_id,
                        a.author
                 FROM books AS b LEFT JOIN authors AS a ON b.author_id = a.author_id),
           rat AS (SELECT book_id,
                          AVG(rating) AS rating_avg
                   FROM ratings 
                   GROUP BY book_id
                   HAVING COUNT(rating_id) >= 50)           
           SELECT b.author,
                  ROUND(AVG(rat.rating_avg), 2) AS rating_avg
           FROM b JOIN rat ON b.book_id = rat.book_id                  
           GROUP BY b.author
           ORDER BY rating_avg DESC
           LIMIT 1;
        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  rating_avg
0  J.K. Rowling/Mary GrandPré        4.28

**J.K. Rowling/Mary GrandPré** имеют самую высокую среднюю оценку книг (4.28), если учитывать только книги с 50 и более оценками.

###  Среднее количество обзоров от пользователей, которые поставили больше 48 оценок

*Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.*

In [10]:
query = '''WITH
           rev AS (SELECT username,
                          COUNT(review_id) AS review_cnt
                          FROM reviews
                          GROUP BY username),
           rat AS (SELECT username,
                          COUNT(rating_id) AS rating_cnt
                          FROM ratings
                          GROUP BY username
                          HAVING COUNT(rating_id) > 48)  
           SELECT AVG(rev.review_cnt) AS avg_review_cnt                 
           FROM rev JOIN rat ON rev.username = rat.username
        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg_review_cnt
0            24.0

В среднем **24 обзора** от пользователей, которые поставили больше 48 оценок.

## Общий вывод

После 1 января 2000 года вышла 821 книга. Есть, что предложить читателям как из классики, так и из современной литературы.  
Для удобства пользователей сформирована сводная таблица с информацией о каждой книге: автор, название, средняя оценка и количество обзоров. Данные представлены по убыванию средней оценки книг.    

Наиболее плодотворное сотрудничество может быть с издательством Penguin Books, которое выпустило 42 книги - максимальное число книг толще 50 страниц.  

J.K. Rowling/Mary GrandPré имеют самую высокую среднюю оценку книг, если учитывать только книги с 50 и более оценками.  
  
Самые активные пользователи пишут обзоры на книги в два раза реже, чем оценивают произведения. В среднем 24 обзора от пользователей, которые поставили больше 48 оценок. 